### PR TITLE
Opt-in to rustc_private for `rust-analyzer`

### PR DESCRIPTION
**Pull request by DJMcNab**
rust-analyzer/rust-analyzer#7891

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -40,3 +40,7 @@ libc = "0.2"
 compiletest_rs = { version = "0.5", features = ["tmp"] }
 rustc_version = "0.2.3"
 colored = "2"
+
+[package.metadata.rust-analyzer]
```
Comment (RalfJung) [inline Cargo.toml:44]: Is there a documentation of `package.metadata.rust-analyzer` keys that we could link to here?

**Comment by RalfJung** [comment]
Thanks. :)

Does this mean I can remove this from my vscode config?
```
"rust-analyzer.rustcSource": "/home/r/.rustup/toolchains/miri/lib/rustlib/rustc-src/rust/compiler/rustc_driver/Cargo.toml"
```

File: Cargo.toml (hunk 1 above)
Comment (DJMcNab) [reply Cargo.toml:44]: The only documentation is that PR, although some more might be coming soon. CC @jyn514

Comment (RalfJung) [reply Cargo.toml:44]: Would be good to have *some* link here, a link to the PR would be a start.

**Comment by DJMcNab** [comment]
You can't remove that unfortunately - this change simply makes the opt-in per-crate explicit.

I did ask @matklad about that, but at the moment that is not the plan. (The order of operations would make that a more intrusive change at the moment, among other reasons). It might be worth having a rust-analyzer issue to track that question though - see [the zulip conversation](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Fixed.20crate.20graphs.20and.20optional.20builtin.20crates/near/229333373)

This means that having that setting set in your vscode config will no longer adversely effect rust-analyzer performance in crates without rustc_private.

However,
```json
{ "rust-analyzer.rustcSource": "discover" }
```
might work to fetch that toolchain

**Comment by RalfJung** [comment]
I saw this "discover" mentioned in your PR but hadn't seen it in the docs before so I was curious about it... I'll have to try that, thanks!

File: Cargo.toml (hunk 1 above)
Comment (DJMcNab) [reply Cargo.toml:44]: Sorry, I missed this comment yesterday

**Comment by RalfJung** [comment]
Thanks. :) @bors r+

**Comment by RalfJung** [comment]
Yes, "discover" works, very nice. :) This should be advertised much better... (or be the default?)

**Comment by DJMcNab** [comment]
Again, there have been discussions about that, but it's non trivial to plumb through. See also https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Fixed.20crate.20graphs.20and.20optional.20builtin.20crates/near/229387042

**Pull request merged**